kernel/environ.mli

```diff
@@ -28,16 +28,60 @@ open Declarations
    - a set of universe constraints
    - a flag telling if Set is, can be, or cannot be set impredicative *)
 
+type lazy_val
+
+val build_lazy_val : lazy_val -> (Vmvalues.values * Id.Set.t) -> unit
+val force_lazy_val : lazy_val -> (Vmvalues.values * Id.Set.t) option
+val dummy_lazy_val : unit -> lazy_val
+
+(** Linking information for the native compiler *)
+type link_info =
+  | Linked of string
+  | LinkedInteractive of string
+  | NotLinked
+
+type key = int CEphemeron.key option ref
+
+type constant_key = constant_body * (link_info ref * key)
+
+type mind_key = mutual_inductive_body * link_info ref
+
+type globals = {
+  env_constants : constant_key Cmap_env.t;
+  env_inductives : mind_key Mindmap_env.t;
+  env_modules : module_body MPmap.t;
+  env_modtypes : module_type_body MPmap.t
+}
+
+type stratification = {
+  env_universes : UGraph.t;
+  env_engagement : engagement
+}
+
+type named_context_val = private {
+  env_named_ctx : Context.Named.t;
+  env_named_map : (Context.Named.Declaration.t * lazy_val) Id.Map.t;
+}
+
+type rel_context_val = private {
+  env_rel_ctx : Context.Rel.t;
+  env_rel_map : (Context.Rel.Declaration.t * lazy_val) Range.t;
+}
+
+type env = {
```

Comment:
I'm not sure that making the `env` type fully transparent is a good idea. We should instead provide a view function from and into this record representation.

Reply:
Ok, but this is orthogonal, since `pre_env` is already transparent today.

Reply:
Yep, but nobody uses `pre_env` whereas `env` is a pervasive, opaque, datastructure.

Reply:
Ok, but isn't making it private enough? What do you want to ensure with the view?

Reply:
Marking it private is a good start, but in general there is no reason to export the implementation of a type if nobody relies on it. That makes changing the representation at a later time harder, and the fact we already changed a few implementation details of environments makes me wary that it ends up having unexpected costs.

Also, it might be enough to check for the direct uses of the fields and provide the corresponding functions in the API.

Reply:
Ok, I'll make it private and re-export the projections, although I'm not convinced it will make the code more robust.